firefox-ios/Client/Frontend/Browser/BrowserPrompts.swift

```diff
@@ -113,3 +113,150 @@ private func titleForJavaScriptPanelInitiatedByFrame(_ frame: WKFrameInfo) -> St
     }
     return title
 }
+
+// MARK: Part of the FXIOS-10334 refactor
+/// This code duplicates existing alerts for a short moment in time to enable having a rollout of JS changes in production
+/// This code is brittle, and if there's any problems this will enable reverting easily
+
+@objc
+protocol NewJSPromptAlertControllerDelegate: AnyObject {
+    func promptAlertControllerDidDismiss(_ alertController: NewJSPromptAlertController, withResult result: Any?)
+}
+
+/// A simple version of UIAlertController that attaches a delegate to the viewDidDisappear method
+/// to allow forwarding the event. The reason this is needed for prompts from Javascript is we
+/// need to invoke the completionHandler passed to us from the WKWebView delegate or else
+/// a runtime exception is thrown.
+class NewJSPromptAlertController: UIAlertController {
+    var alertInfo: NewJSAlertInfo?
+    weak var delegate: NewJSPromptAlertControllerDelegate?
+    private var handledAction: Bool = false
+    private var dismissalResult: Any?
+
+    convenience init(
+        title: String?,
+        message: String?,
+        preferredStyle: UIAlertController.Style = .alert,
+        alertInfo: NewJSAlertInfo
+    ) {
+        self.init(title: title, message: message, preferredStyle: preferredStyle)
+        self.alertInfo = alertInfo
+    }
+
+    /// Set the result to pass during dismissal
+    func setDismissalResult(_ result: Any?) {
+        handledAction = true
+        dismissalResult = result
+    }
+
+    override func viewDidDisappear(_ animated: Bool) {
+        super.viewDidDisappear(animated)
+
+        // The alert controller is dismissed before the UIAlertAction handler is called so a delay is needed
+        DispatchQueue.main.asyncAfter(deadline: .now() + 0.3) { [self] in
+            // Call the dismissal completion handler if it wasn't handled yet
+            if !handledAction {
+                alertInfo?.cancel()
+            }
+
+            // Notify the delegate about dismissal and pass the result
+            delegate?.promptAlertControllerDidDismiss(self, withResult: dismissalResult)
+        }
+    }
```

Comment:
@lmarceau Why do we need this async dispatch at all? Shouldn't we be able to rely on the action completion handlers + other related callbacks entirely? I could be misunderstanding the intent, just wanted to clarify.

Reply:
If we don't have it then for example with the `NewMessageAlert` the following code gets called after the `viewDidDisappear` is called in the ViewController.
```
            UIAlertAction(title: .OKString, style: .default) { [weak alertController] _ in
                alertController?.setDismissalResult(nil)
                self.completionHandler()
            }
```
This means the `completionHandler` would be called twice (one time from the alert handler, and one time since the `handledAction` property would still be false).

Not going to lie, I totally picked this up from [Brave code](https://github.com/brave/brave-core/blob/7d3a262c081b59e81f610d2da0beb90199fc6161/ios/brave-ios/Sources/Brave/Frontend/Browser/BrowserPrompts.swift#L70) when I was having issues getting the alerts properly dismissed on all cases 🤫